### ISSUE TITLE
`unused_io_amount` fails to fire on offending code

### ISSUE THREAD

**Issue by bugadani** [rust-lang/rust-clippy#8554]
### Summary

Code on playground:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=3deb8813d6956a51bdffe787d721c7b8

Code was found in a bigger codebase where the same lint previously fired on some other patterns. The example is obviously incorrect as it does not care about how many bytes were written in any of the `write` calls. Considering how close this code is to the documented example, I'm not entirely sure whether it's an actual bug or a PEBKAC.

### Lint Name

unused_io_amount

### Reproducer

I tried this code:

```rust
use std::io::Write;

const END: u8 = 0xC0;
const ESC: u8 = 0xDB;

const ESC_END: u8 = 0xDC;
const ESC_ESC: u8 = 0xDD;

pub fn encode_byte(writer: &mut impl Write, byte: u8) -> std::io::Result<()> {
    match byte {
        END => writer.write(&[ESC, ESC_END])?,
        ESC => writer.write(&[ESC, ESC_ESC])?,
        byte => writer.write(&[byte])?,
    };

    Ok(())
}
```

I expected to see this happen:

Clippy should have reported errors.

Instead, this happened:

Nothing.

### Version

```text
rustc 1.61.0-nightly (335ffbfa5 2022-03-11)
binary: rustc
commit-hash: 335ffbfa547df94ac236f5c56130cecf99c8d82b
commit-date: 2022-03-11
host: x86_64-pc-windows-msvc
release: 1.61.0-nightly
LLVM version: 14.0.0
```
